specs/NestedFrame.md

```diff
@@ -0,0 +1,215 @@
+main page and first-level iframes (the default behavior), a
+partial WebView2 frames tree with specific iframes of interest,
+or the full WebView2 frames tree.
+
+# Examples
+### C++ Sample
+```cpp
+wil::com_ptr<ICoreWebView2> m_webview;
+std::map<int, std::vector<std::wstring>> m_frame_navigation_urls;
+// In this example, a WebView2 application wants to manage the 
+// navigation of third-party content residing in second-level iframes
+// (Main frame -> First-level frame -> Second-level third-party frames).
+HRESULT RecordThirdPartyFrameNavigation() {
+    auto webview2_4 = m_webView.try_query<ICoreWebView2_4>();
+    // Track the first-level webview frame.
+    webview2_4->add_FrameCreated(
+            Callback<ICoreWebView2FrameCreatedEventHandler>(
+                [this](ICoreWebView2* sender, ICoreWebView2FrameCreatedEventArgs* args)
+                    -> HRESULT {
+                    // [AddFrameCreated]
+                    wil::com_ptr<ICoreWebView2Frame> webviewFrame;
+                    CHECK_FAILURE(args->get_Frame(&webviewFrame));
+                    // Track nested (second-level) webview frame.
+                    auto frame7 = webviewFrame.try_query<ICoreWebView2Frame7>();
+                    frame7->add_FrameCreated(
+                        Callback<ICoreWebView2FrameChildFrameCreatedEventHandler>(
+                            [this](
+                                ICoreWebView2Frame* sender,
+                                ICoreWebView2FrameCreatedEventArgs* args) -> HRESULT
+                            {
+                                wil::com_ptr<ICoreWebView2Frame> webviewFrame;
+                                CHECK_FAILURE(args->get_Frame(&webviewFrame));
+                                wil::com_ptr<ICoreWebView2Frame2> frame2 = 
+                                    webviewFrame.try_query<ICoreWebView2Frame2>();
+                                if (frame2)
+                                {
+                                    // Subscribe to nested (second-level) webview frame navigation 
+                                    // starting event.
+                                    frame2->add_NavigationStarting(
+                                        Callback<ICoreWebView2FrameNavigationStartingEventHandler>(
+                                            [this](
+                                                ICoreWebView2Frame* sender,
+                                                ICoreWebView2NavigationStartingEventArgs* args) -> HRESULT {
+                                                // Manage the navigation, e.g. cancel the 
+                                                // navigation if it's on block list.
+                                                UINT32 frameId = 0;
+                                                auto frame5 = wil::com_ptr<ICoreWebView2Frame>(sender)
+                                                                .try_query<ICoreWebView2Frame5>();

```

Comment:
Missing null check from try_query. Maybe we should use ICoreWebView2Frame5 in the outer lambda, so that we don't even get here if v5 is not supported. Then that would allow us to use query() instead of try_query().

Reply:
Please consider simple fix of null check for simple sample code.